File: Swift/프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.swift
import Foundation

func solution(_ numbers:[Int], _ hand:String) -> String {
    let keypad: [Int: [Int]] = [
        1: [0, 0], 2: [0, 1], 3: [0, 2],
        4: [1, 0], 5: [1, 1], 6: [1, 2],
        7: [2, 0], 8: [2, 1], 9: [2, 2],
        0: [3, 1]
    ]
    let leftHandN = [1, 4, 7]
    let rightHandN = [3, 6, 9]
    
    var leftPos = [3, 0]
    var rightPos = [3, 2] 
    
    var result = ""
    for number in numbers {
        if leftHandN.contains(number) {
            result += "L"
            leftPos = keypad[number]!
        } else if rightHandN.contains(number) {
            result += "R"
            rightPos = keypad[number]!
        } else {
            let currentPos = keypad[number]!
            
            let leftDist = abs(leftPos[0] - currentPos[0]) + abs(leftPos[1] - currentPos[1])
            let rightDist = abs(rightPos[0] - currentPos[0]) + abs(rightPos[1] - currentPos[1])
            
            if leftDist == rightDist {
                if hand == "left" {
                    result += "L"
                    leftPos = currentPos
                } else {
                    result += "R"
                    rightPos = currentPos
                }
                continue
            }
            
            if leftDist < rightDist {
                result += "L"
                leftPos = currentPos
            } else {
                result += "R"
                rightPos = currentPos
            }
        }
    }
    
    return result
}

File: Java/프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.*;

class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        
        int[][] keypad = {
            {3, 1},
            {0, 0},
            {0, 1},
            {0, 2},
            {1, 0},
            {1, 1},
            {1, 2},
            {2, 0},
            {2, 1},
            {2, 2},
        };
        
        int[] leftPos = {3, 0};
        int[] rightPos = {3, 2};
        
        for (int num : numbers) {
            if (num == 1 || num == 4 || num == 7) {
                sb.append("L");
                leftPos = keypad[num];
            } else if (num == 3 || num == 6 || num == 9) {
                sb.append("R");
                rightPos = keypad[num];
            } else {
                int[] keypadPos = keypad[num]; 
                int lDist = getDistance(leftPos, keypadPos);
                int rDist = getDistance(rightPos, keypadPos);
                
                if (lDist < rDist) {
                    sb.append("L");
                    leftPos = keypad[num];
                } else if (lDist > rDist) {
                    sb.append("R");
                    rightPos = keypad[num];
                } else {
                    if (hand.equals("left")) {
                        sb.append("L");
                        leftPos = keypad[num];
                    } else {
                        sb.append("R");
                        rightPos = keypad[num];
                    }
                }
            }
        }
        return sb.toString();
    }
    
    static private int getDistance(int[] position, int[] target) {
        return Math.abs(position[0] - target[0]) + Math.abs(position[1] - target[1]);
    }
}